fendermint/vm/resolver/src/ipld.rs

```diff
@@ -0,0 +1,131 @@
+// Copyright 2022-2023 Protocol Labs
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use std::{future::Future, time::Duration};
+
+use async_stm::{atomically, queues::TQueueLike};
+use ipc_ipld_resolver::Client as ResolverClient;
+use ipc_sdk::subnet_id::SubnetID;
+
+use crate::pool::{ResolveQueue, ResolveTask};
+
+/// The IPLD Resolver takes resolution tasks from the [ResolvePool] and
+/// uses the [ipc_ipld_resolver] to fetch the content from subnets.
+pub struct IpldResolver {
+    client: ResolverClient,
+    queue: ResolveQueue,
+    retry_delay: Duration,
+    own_subnet_id: SubnetID,
+}
+
+impl IpldResolver {
+    pub fn new(
+        client: ResolverClient,
+        queue: ResolveQueue,
+        retry_delay: Duration,
+        own_subnet_id: SubnetID,
+    ) -> Self {
+        Self {
+            client,
+            queue,
+            retry_delay,
+            own_subnet_id,
+        }
+    }
+
+    /// Start taking tasks from the resolver pool and resolving them using the IPLD Resolver.
+    pub async fn run(self) {
+        loop {
+            let (task, use_own_subnet) = atomically(|| {
+                let task = self.queue.read()?;
```

Comment:
Good question, yes and no. The original version of STM would block this thread, but the async version doesn't. It will just add itself to the wait queue of the `TVar` and `await` for some change. Because it's `async` and using `tokio`, it will not block, but it also won't return until something appears in the queue.